File: erlang/raindrops/src/raindrops.erl
-module(raindrops).

-export([convert/1]).

-define(RAINDROPS, [
    {3, "Pling"},
    {5, "Plang"},
    {7, "Plong"}
]).

-spec convert(Number :: non_neg_integer()) -> string().
convert(Number) ->
    do_convert(Number, [], ?RAINDROPS).

-spec do_convert(
    Number :: non_neg_integer(),
    Sounds :: [string()],
    Raindrops :: [{non_neg_integer(), string()}]
) ->
    string().
do_convert(Number, [], []) ->
    integer_to_list(Number);
do_convert(_, Sounds, []) ->
    lists:concat(lists:reverse(Sounds));
do_convert(Number, Sounds, [{Divisor, Sound} | Raindrops]) when
    Number rem Divisor =:= 0
->
    do_convert(Number, [Sound | Sounds], Raindrops);
do_convert(Number, Sounds, [_ | Raindrops]) ->
    do_convert(Number, Sounds, Raindrops).

File: elixir/raindrops/lib/raindrops.ex
defmodule Raindrops do
  @doc """
  Returns a string based on raindrop factors.

  - If the number contains 3 as a prime factor, output 'Pling'.
  - If the number contains 5 as a prime factor, output 'Plang'.
  - If the number contains 7 as a prime factor, output 'Plong'.
  - If the number does not contain 3, 5, or 7 as a prime factor,
    just pass the number's digits straight through.
  """

  @raindrops [
    {3, "Pling"},
    {5, "Plang"},
    {7, "Plong"}
  ]

  @spec convert(pos_integer) :: String.t()
  def convert(number) do
    do_convert(number, [], @raindrops)
  end

  defp do_convert(number, [], []), do: Integer.to_string(number)

  defp do_convert(_, sounds, []) do
    sounds
    |> Enum.reverse()
    |> Enum.join("")
  end

  defp do_convert(number, sounds, [{divisor, sound} | raindrops])
       when rem(number, divisor) == 0 do
    do_convert(number, [sound | sounds], raindrops)
  end

  defp do_convert(number, sounds, [_ | raindrops]) do
    do_convert(number, sounds, raindrops)
  end
end
